src/device_trezor/trezor/transport.cpp

```diff
@@ -0,0 +1,672 @@
+      MERROR("Message could not be serialized");
+      throw exc::EncodingException("Message could not be serialized");
+    }
+
+    std::string uri = "/call/" + m_session.get();
+    std::string req_hex = epee::to_hex::string(epee::span<const std::uint8_t>(req_buff_raw, buff_size));
+    std::string res_hex;
+
+    bool req_status = invoke_bridge_http(uri, req_hex, res_hex, m_http_client);
+    if (!req_status){
+      MERROR("Call method failed");
+      throw exc::CommunicationException("Call method failed");
+    }
+
+    m_response = res_hex;
+    return true;
+  }
+
+  bool BridgeTransport::read(std::shared_ptr<google::protobuf::Message> & msg, messages::MessageType * msg_type) {
+    if (!m_response){
+      MERROR("Could not read, no response stored");
+      return false;
+    }
+
+    std::string bin_data;
+    if (!epee::string_tools::parse_hexstr_to_binbuff(m_response.get(), bin_data)){
+      MERROR("Response is not well hexcoded");
+      throw exc::CommunicationException("Response is not well hexcoded");
+    }
+
+    uint16_t msg_tag;
+    uint32_t msg_len;
+    deserialize_message_header(bin_data.c_str(), msg_tag, msg_len);
+    if (bin_data.size() != msg_len + 6){
+      MERROR("Response size invalid");
+      throw exc::CommunicationException("Response is not well hexcoded");
+    }
+
+    if (msg_type){
+      *msg_type = static_cast<messages::MessageType>(msg_tag);
+    }
+
+    std::shared_ptr<google::protobuf::Message> msg_wrap(MessageMapper::get_message(msg_tag));
+    if (!msg_wrap->ParseFromArray(bin_data.c_str() + 6, msg_len)){
+      MERROR("Message could not be parsed");
+      throw exc::EncodingException("Response is not well hexcoded");
+    }
+    msg = msg_wrap;
+    return true;
+  }
+
+  const boost::optional<json_val> & BridgeTransport::device_info() const {
+    return m_device_info;
+  }
+
+  std::ostream& BridgeTransport::dump(std::ostream& o) const {
+    return o << "BridgeTransport<path=" << (m_device_path ? m_device_path.get() : "None")
+             << ", info=" << (m_device_info ? t_serialize(m_device_info.get()) : "None")
+             << ", session=" << (m_session ? m_session.get() : "None")
+             << ">";
+  }
+
+  //
+  // UdpTransport
+  //
+  const char * UdpTransport::PATH_PREFIX = "udp:";
+  const char * UdpTransport::DEFAULT_HOST = "127.0.0.1";
+  const int UdpTransport::DEFAULT_PORT = 21324;
+
+  UdpTransport::UdpTransport(boost::optional<std::string> device_path,
+                             boost::optional<std::shared_ptr<Protocol>> proto) :
+      m_io_service(), m_deadline(m_io_service)
+  {
+    m_device_port = DEFAULT_PORT;
+    if (device_path) {
+      const std::string device_str = device_path.get();
+      auto delim = device_str.find(":");
+      if (delim == std::string::npos) {
+        m_device_host = device_str;
+      } else {
+        m_device_host = device_str.substr(0, delim);
+        m_device_port = std::stoi(device_str.substr(delim + 1));
```

Comment:
Does not check for errors (though maybe it's been checked already elsewhere).

Reply:
I've added additional check for the port number range but if the stoi itself fails I am fine with it throwing illegal argument exception